File: docker/ai-service/routers/analyze_email.py
"""Orbit — routeur POST /analyze-email (extraction de rendez-vous).

Parité avec le mini-service bun (src/routers/analyze-email.ts) :
- garde-fou déterministe sur les participants (uniquement les adresses
  réellement citées dans l'email source, expéditeur inclus) ;
- correction d'année (un RDV email est toujours futur) ;
- cache mémoire 10 min (clé = hash du payload, « now » tronqué à l'heure).
Les bornes métier finales restent côté Next.js (suggestionFromExtracted).
"""

import re
from typing import List, Optional

from fastapi import APIRouter
from fastapi.responses import JSONResponse

from models.schemas import AnalyzeEmailRequest
from services.cache import cache_get, cache_set
from services.ollama_client import default_client
from services.prompts import TEMPLATES, render
from services.utils import (
    LIMITS,
    fnv1a_hex,
    format_fr_long,
    hour_bucket,
    parse_json_loose,
    parse_now,
    sanitize,
)

router = APIRouter()

_EMAIL_RE = re.compile(r"[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}")


def _result_from_json(parsed: dict, now, subject: str, from_: str, body: str) -> dict:
    # ── Participants : l'IA peut inventer des adresses — on ne conserve QUE
    # les adresses réellement présentes dans l'email source (regex + filtre).
    source_lower = f"{subject}\n{from_}\n{body}".lower()
    input_emails = list(dict.fromkeys(_EMAIL_RE.findall(source_lower)))
    ai_attendees = [
        a.strip()[:254]
        for a in (parsed.get("attendees") or [])
        if isinstance(a, str) and "@" in a and a.strip().lower() in source_lower
    ]
    attendees = list(dict.fromkeys(input_emails + ai_attendees))[:10]

    # ── Année : un rendez-vous email est TOUJOURS futur.
    start_time = parsed.get("startTime") if isinstance(parsed.get("startTime"), str) else ""
    end_time = parsed.get("endTime") if isinstance(parsed.get("endTime"), str) else ""
    year_adjusted = False
    try:
        from datetime import datetime

        start_dt = datetime.fromisoformat(start_time.replace("Z", "+00:00"))
        delta_days = (now - start_dt).total_seconds() / 86400.0
        if delta_days > 90:
            candidate = start_dt.replace(year=now.year)
            if (candidate - now).total_seconds() > -86400.0:
                shift = (candidate - start_dt).total_seconds()
                start_time = candidate.isoformat()
                try:
                    end_dt = datetime.fromisoformat(end_time.replace("Z", "+00:00"))
                    if abs((end_dt - start_dt).total_seconds()) < 48 * 3600:
                        from datetime import timedelta

                        end_time = (end_dt + timedelta(seconds=shift)).isoformat()
                except ValueError:
                    pass
                year_adjusted = True
    except ValueError:
        pass

    location = parsed.get("location") if isinstance(parsed.get("location"), str) else None
    result = {
        "isEvent": parsed.get("isEvent") is not False,
        "title": (parsed.get("title") or "").strip()[:120] if isinstance(parsed.get("title"), str) else "",
        "description": (parsed.get("description") or "")[:500]
        if isinstance(parsed.get("description"), str)
        else "",
        "startTime": start_time,
        "endTime": end_time,
        "durationMinutes": parsed.get("durationMinutes") or 60,
        "location": location.strip()[:200] if location and location.strip() else None,
        "attendees": attendees,
        "confidence": parsed.get("confidence") or 0.5,
    }
    if year_adjusted:
        result["yearAdjusted"] = True
    return result


@router.post("/analyze-email")
async def analyze_email(req: AnalyzeEmailRequest) -> JSONResponse:
    subject = sanitize(req.subject, LIMITS["subject"])
    from_ = sanitize(req.from_, 320)
    body = sanitize(req.bodyText, LIMITS["emailBody"])
    if not subject and not body:
        return JSONResponse({"error": "subject ou bodyText requis"}, status_code=400)

    now = parse_now(req.now)
    timezone = sanitize(req.timezone, 64) or "Africa/Dakar"

    # ── Cache (payload utile + heure courante) ───────────────────────────────
    cache_key = "analyze:" + fnv1a_hex(
        f"{subject}|{from_}|{body}|{hour_bucket(now)}|{timezone}"
    )
    cached = cache_get(cache_key)
    if cached is not None:
        return JSONResponse({**cached, "cached": True})

    # ── Prompt système (template + interpolation) ────────────────────────────
    system = render(
        TEMPLATES["email_analysis"],
        NOW_FR=format_fr_long(now),
        NOW_ISO=now.isoformat(),
        TIMEZONE=timezone,
    )
    user = f"Objet : {subject}\nDe : {from_}\n\nCorps de l'email :\n{body}"

    # ── Inférence : Ollama en mode JSON forcé ────────────────────────────────
    client = default_client()
    data = await client.generate(user, system, json_mode=True)
    if data is None:
        return JSONResponse(
            {
                "error": f"Ollama indisponible ({client.base_url}) — vérifiez que le modèle "
                f"{client.model} est téléchargé (ollama pull {client.model})"
            },
            status_code=502,
        )

    content = data.get("response", "")
    parsed = parse_json_loose(content)
    if parsed is None:
        return JSONResponse(
            {
                "result": None,
                "provider": "ollama",
                "message": "Réponse non exploitable",
                "rawResponse": content[: LIMITS["rawResponse"]],
            }
        )

    result = _result_from_json(parsed, now, subject, from_, body)
    payload = {
        "result": result,
        "provider": "ollama",
        # Debug côté client (tronqué) — jamais loggé côté service.
        "rawResponse": content[: LIMITS["rawResponse"]],
    }
    cache_set(cache_key, payload)
    return JSONResponse(payload)

File: mini-services/ai-service/src/routers/analyze-email.ts
// ═══════════════════════════════════════════════════════════════════════════
// Orbit — Micro-service IA · routeur POST /analyze-email
// ───────────────────────────────────────────────────────────────────────────
// Extraction de rendez-vous depuis un email : date, horaires, durée, LIEU et
// PARTICIPANTS (enrichi Prompt IA locale). Cache mémoire 10 min (clé = hash
// du payload, « now » tronqué à l'heure). Les bornes métier finales restent
// côté Next.js (suggestionFromExtracted).
// ═══════════════════════════════════════════════════════════════════════════

import { LIMITS } from "../config"
import { cacheGet, cacheSet } from "../cache"
import { render, templates } from "../prompts"
import { completeJsonRobust } from "../services/llm"
import type { AnalyzeEmailBody } from "../models/schemas"
import { formatFrLong, hashString, hourBucket, sanitize } from "../utils"

export async function handleAnalyzeEmail(body: AnalyzeEmailBody): Promise<Response> {
  const subject = sanitize(body.subject, LIMITS.subject)
  const from = sanitize(body.from, 320)
  const bodyText = sanitize(body.bodyText, LIMITS.emailBody)
  if (!subject && !bodyText) {
    return Response.json({ error: "subject ou bodyText requis" }, { status: 400 })
  }

  const now = body.now && !Number.isNaN(Date.parse(body.now)) ? new Date(body.now) : new Date()
  const timezone = sanitize(body.timezone, 64) || "Africa/Dakar"

  // ── Cache (payload utile + heure courante) ───────────────────────────────
  const cacheKey = `analyze:${hashString(
    JSON.stringify({ subject, from, bodyText, h: hourBucket(now.toISOString()), timezone })
  )}`
  const cached = cacheGet<unknown>(cacheKey)
  if (cached) {
    return Response.json({ ...(cached as Record<string, unknown>), cached: true })
  }

  // ── Prompt système (template + interpolation) ────────────────────────────
  const system = render(templates.emailAnalysis, {
    NOW_FR: formatFrLong(now),
    NOW_ISO: now.toISOString(),
    TIMEZONE: timezone,
  })
  const user = [
    `Objet : ${subject}`,
    `De : ${from}`,
    "",
    "Corps de l'email :",
    bodyText,
  ].join("\n")

  // ── Inférence : Ollama (JSON forcé) → fallback SDK ───────────────────────
  const completion = await completeJsonRobust(
    system,
    user,
    '{"isEvent": true|false, "title": "titre court", "description": "résumé 1-2 phrases", "startTime": "ISO 8601", "endTime": "ISO 8601", "durationMinutes": 60, "location": "lieu ou null", "attendees": ["emails"], "confidence": 0.9}'
  )
  if (!completion) {
    return Response.json({ error: "Aucune réponse du moteur IA" }, { status: 502 })
  }
  const { json, content, provider, repaired } = completion

  if (!json) {
    return Response.json({
      result: null,
      provider,
      message: "Réponse non exploitable",
      rawResponse: content.slice(0, LIMITS.rawResponse),
    })
  }

  // ── Normalisation douce + garde-fous déterministes ───────────────────────
  // Participants : l'IA peut inventer des adresses — on ne conserve QUE les
  // adresses réellement présentes dans l'email source, en y ajoutant celles
  // détectées par regex (expéditeur inclus). Aucune invention ne passe.
  const sourceLower = `${subject}\n${from}\n${bodyText}`.toLowerCase()
  const inputEmails = [...new Set(sourceLower.match(/[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}/g) ?? [])]
  const aiAttendees = (Array.isArray(json.attendees) ? json.attendees : [])
    .filter((a): a is string => typeof a === "string")
    .map((a) => a.trim().slice(0, 254))
    .filter((a) => a.includes("@") && sourceLower.includes(a.toLowerCase()))
  const attendees = [...new Set([...inputEmails, ...aiAttendees])].slice(0, 10)

  // Année : un rendez-vous email est TOUJOURS dans le futur — si l'IA renvoie
  // une date à plus de 90 jours dans le passé alors que la même date ramenée
  // à l'année courante est future (email sans année explicite), on corrige.
  let startTime = typeof json.startTime === "string" ? json.startTime : ""
  let endTime = typeof json.endTime === "string" ? json.endTime : ""
  let yearAdjusted = false
  const startMs = Date.parse(startTime)
  if (!Number.isNaN(startMs) && now.getTime() - startMs > 90 * 86_400_000) {
    const candidate = new Date(startMs)
    candidate.setUTCFullYear(now.getUTCFullYear())
    if (candidate.getTime() > now.getTime() - 86_400_000) {
      const shift = candidate.getTime() - startMs
      startTime = candidate.toISOString()
      const endMs = Date.parse(endTime)
      // La fin suit le même décalage si elle reste collée au début (< 48 h).
      if (!Number.isNaN(endMs) && Math.abs(endMs - startMs) < 48 * 3_600_000) {
        endTime = new Date(endMs + shift).toISOString()
      }
      yearAdjusted = true
    }
  }

  const result = {
    isEvent: json.isEvent !== false,
    title: typeof json.title === "string" ? json.title.trim().slice(0, 120) : "",
    description: typeof json.description === "string" ? json.description.slice(0, 500) : "",
    startTime,
    endTime,
    durationMinutes: Number(json.durationMinutes) || 60,
    location: typeof json.location === "string" && json.location.trim() ? json.location.trim().slice(0, 200) : null,
    attendees,
    confidence: Number(json.confidence) || 0.5,
    ...(yearAdjusted ? { yearAdjusted: true } : {}),
  }

  const payload = {
    result,
    provider,
    ...(repaired ? { repaired: true } : {}),
    // Debug côté client (tronqué) — jamais loggé côté service (confidentialité).
    rawResponse: content.slice(0, LIMITS.rawResponse),
  }
  cacheSet(cacheKey, payload)
  return Response.json(payload)
}
